pkg/ingress/ingress.go

```diff
@@ -100,22 +99,33 @@ func ExpandedHosts(hosts sets.String) sets.String {
 	expanded := make(sets.String, len(hosts)*len(allowedSuffixes))
 	for _, h := range hosts.List() {
 		for _, suffix := range allowedSuffixes {
-			if strings.HasSuffix(h, suffix) {
-				trimmedHost := strings.TrimSuffix(h, suffix)
-				if isValidTopLevelDomain(trimmedHost) {
-					expanded.Insert(trimmedHost)
+			if th := strings.TrimSuffix(h, suffix); suffix == "" || len(th) < len(h) {
+				if isValidTopLevelDomain(th) {
+					expanded.Insert(th)
 				}
 			}
 		}
 	}
 	return expanded
 }
 
-// Validate that the Top Level Domain of a given hostname is not fully numeric.
+// Validate that the Top Level Domain of a given hostname is valid.
+// Current checks:
+//  - not all digits
+//  - len < 64
 // Example: '1234' is an invalid TLD
 func isValidTopLevelDomain(domain string) bool {
 	parts := strings.Split(domain, ".")
-	topLevelDomain := parts[len(parts)-1]
-	_, err := strconv.Atoi(topLevelDomain)
-	return (err != nil)
+	tld := parts[len(parts)-1]
+	if len(tld) > 63 {
+		return false
+	}
+	for _, c := range []byte(tld) {
+		c -= '0'
+		if c > 9 {
```

Comment:
Do all allowed characters in a domain have a byte value larger than that of '0'?

Reply:
hmmm, interesting, `-` I think before numbers. let's do a more precise check.

Reply:
So, `-` is smaller than `0`, so I fixed that. Added a test.
Also improved how we generate test failures.